src/v/transform/api.cc

```diff
@@ -83,10 +87,15 @@ class partition_source final : public source {
 
     ss::future<model::record_batch_reader>
     read_batch(kafka::offset offset, ss::abort_source* as) final {
+        auto _ = _gate.hold();
         // There currently no way to abort the call to get the sync start, so
         // instead we wrap the resulting future in our abort source.
         auto result = co_await ssx::with_timeout_abortable(
-          _partition.sync_effective_start(), model::no_timeout, *as);
+          // Ensure we don't delete the partition until this has resolved if we
+          // end up timing out.
+          _partition.sync_effective_start().finally([_ = _gate.hold()] {}),
```

Comment:
this is a curious one. i don't wanna say that we don't have cases of constructing a capture list containing a direct call to `gate.hold()` (plenty where we move the handle in), but it is making me wonder what happens here if the gate is closed.

I think `_partition.sync_effective_start()` would still create a future--well it must, since it is what `.finally` is attached to. but then finally will never run because its parameter (the lambda) wouldn't be able to be created since hold would throw.

the entire expression would throw, but I would bet that the `_partition.sync_effective_start()` would continue to live on as a background future.

Reply:
> but it is making me wonder what happens here if the gate is closed.

The gate is already held open at the start of calling this function

Reply:
oh good point. then you could do `[_ = std::move(_)]` haha

Reply:
Is it more clear if I hold the gate open twice in this function then move one of the holders into the function? I guess holders are also copyable, so I could copy it in...

Reply:
> oh good point. then you could do [_ = std::move(_)] haha it looks so funny with the underscores.

I miss discard variables that are around in other languages, but yeah we'd probably want to copy it? IDK right now we never support calling anything else or having open calls while stop is called anyways, so it's really only guarding against the case when the abort fires and the future is still hanging out in the background.

Reply:
I ended up implementing moving the gate into the `finally` block. It looks sus a bit because we're not holding the gate open in the continuation, but the contract for the processor is for pending futures to resolve before calling stop.